### ISSUE TITLE
Request returning optimum value when trial is pruned

### ISSUE THREAD

**Issue by Tribleave** [optuna/optuna#3168]
When trial is pruned, current strategy select the last one of intermediate values.

However in most scenario, we can get the best model before pruned. So the optimum value of them should be reported to the study.

https://github.com/optuna/optuna/blob/cf03adeedc59013661e13fd8ed3763a7ac00ae8d/optuna/study/study.py#L643-L650

**Comment by nzw0301** [comment]
Thank you for sending your feature request! 

In my understanding, black-box optimisation aims to obtain the best-performed parameters, so we do not care about pruned trials' best value since such trials poorly perform. In addition, we can access intermediate values after optimisation. So I suppose the current implementation has no issue. Could you explain a little bit more about your scenario? 

From the perspective of samplers, pruned trials' returned values can be used to suggest the next trials as mentioned in  
https://github.com/optuna/optuna/issues/1647. So if this feature can improve the sampler's performance, it's worth discussing this feature, I think.

**Comment by Tribleave** [comment]
Thanks you for quick response!

I'm using optuna to find best hyperparameters for **neural networks** with pytorch integration in fact. Tuning that has 2 major differences to traditional algorithms:

1. The metric of the training process could drop at the last few epochs. But the total epochs are based on hyperparameters.
2. It's time-consuming. I early-stop the training by `optuna.pruners.PatientPruner`. 

The training process is pruned once have metrics drop (with some patience).

I have already selected the optimum from intermediate values if the trial completed. It makes sense because I can just get the best metric by stopping training at that step. And I think the same option should be added when a trial is pruned.

**Comment by nzw0301** [comment]
Thank you for your explanation! Indeed, I did a similar thing in my project. Concretely, I retuned the best intermediate value instead of `raise optuna.TrialPruned`.

```python
        if trial.should_prune():
            return min(trial.intermediate_values)
            # raise optuna.TrialPruned
```

According to your comment, you use some integration, it might be impossible to do that.
